designs/CentrallyManagingNuGetPackageVersions-VSManagePackages.md

```diff
@@ -0,0 +1,414 @@
+  </ItemGroup>
+
+</Project>
+```
+
+## Requirements
+
+The requirements are defined in the [feature spec](https://github.com/NuGet/Home/wiki/Centrally-managing-NuGet-package-versions).
+
+Note that:
+
+* It's valid for some, not all, projects in the solution to use centrally managed package versions.
+* It's valid for different projects using central package versions to use different `Directory.Packages.props` files.
+* As is valid now, it's valid for some projects to use `packages.config`, while others use `PackageReference`. The `packages.config` projects cannot use centrally managed package versions, and individual `PackageReference` projects may opt in or out of central package version management.
+
+## Solution
+
+A new service should be created. This service will have multiple responsibilities:
+
+* Write changes to `Directory.Packages.props` files to disk.
+* Find which `Directory.Packages.props` file needs to be changed given a package install for a specific project.
+* Orchestrate efficient updating and nomination of projects. In particular, when project files and `Directory.Packages.props` files need to change due to a single customer gesture, avoid multiple project nominations and multiple restores due to multiple file changes.
+
+This new service will by used by NuGet, but only for projects that have centrally managed versions enabled. Projects that do not use CPVM will continue to work as they do today.
+
+Proposal: When a project is using `CPVM`, NuGet will only use this new service and will no longer call the existing project system APIs
+
+Proposed API. Very subject to change as we start implementing and discover better ideas:
+
+```cs
+interface IDirectoryPackagesPropsService
+{
+  /// <summary>
+  /// Add a PackageReference to the projects provided. If the packageId does
+  /// </summary>
+  Task<bool> AddPackageAsync(List<string> projects, string packageId, string packageVersion);
+
+  /// <summary>
+  /// Remove a PackageReference from the projects listed, but do not remove the PackageVersion
+  /// from the Directory.Packages.props file.
+  /// </summary>
+  Task<bool> RemovePackageAsync(List<string> projects, string packageId);
+
+  /// <summary>
+  /// </summary>
+  Task<bool> UpdatePackageAsync(List<string> projects, string packageId, string packageVersion);
+
+  /// <summary>
+  /// Delete the PackageVersion items from any Directory.Packages.props file imported by the
+  /// provided projects. If any of the projects contain a PackageReference to the project ID,
+  /// throw an error.
+  /// </summary>
+  /// <remarks>
+  /// Out of scope of the first version (minimal viable product). In scope later. Adding it now
+  // avoids changing the interface in shipped versions.
+  /// </remarks>
+  Task<bool> DeletePackageAsync(List<string> projects, string packageId);
+}
+```
+
+Open questions:
+
+* How to handle when a PackageReference needs `IncludeAssets`, `PrivateAssets` and/or `ExcludeAssets`?
```

Comment:
This experience should not change. 
Is the ability to IncludeAssets, etc. is supported in Visual Studio only for SDK style projects?

Reply:
This spec is about the API contract between NuGet and the project systems, so we should define it. I should extend appendix 1 with more details, but basically the way it currently works is we ask project system for an object that represents the PackageReference item in MSBuild, then we call additional APIs to add the extra information. If NuGet is going to call a new VS service to make modifications, the API for the service needs to allow this scenario.